blockprod/src/detail/utils.rs

```diff
@@ -0,0 +1,185 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::BTreeMap;
+
+use chainstate::{
+    chainstate_interface::ChainstateInterface, BlockIndex, GenBlockIndex, NonZeroPoolBalances,
+};
+use chainstate_types::{pos_randomness::PoSRandomness, EpochData};
+use common::{
+    chain::{
+        block::timestamp::{BlockTimestamp, BlockTimestampInternalType},
+        ChainConfig, PoolId,
+    },
+    primitives::{Amount, BlockHeight},
+};
+
+use crate::BlockProductionError;
+
+#[derive(thiserror::Error, Debug, PartialEq, Eq)]
+pub enum PoSAccountingError {
+    #[error("Staker balance retrieval error: {0}")]
+    StakerBalanceRetrievalError(String),
+}
+
+pub fn get_pool_staker_balance(
+    chainstate: &dyn ChainstateInterface,
+    pool_id: &PoolId,
+) -> Result<Amount, BlockProductionError> {
+    let balance = chainstate
+        .get_stake_pool_data(*pool_id)
+        .map_err(|err| {
+            BlockProductionError::ChainstateError(
+                consensus::ChainstateError::StakePoolDataReadError(*pool_id, err.to_string()),
+            )
+        })?
+        .ok_or(BlockProductionError::PoolDataNotFound(*pool_id))?
+        .staker_balance()
+        .map_err(|err| PoSAccountingError::StakerBalanceRetrievalError(err.to_string()))?;
+
+    Ok(balance)
+}
+
+pub fn get_pool_total_balance(
+    chainstate: &dyn ChainstateInterface,
+    pool_id: &PoolId,
+) -> Result<Amount, BlockProductionError> {
+    let pool_balance = chainstate
+        .get_stake_pool_balance(*pool_id)
+        .map_err(|err| {
+            BlockProductionError::ChainstateError(consensus::ChainstateError::PoolBalanceReadError(
+                *pool_id,
+                err.to_string(),
+            ))
+        })?
+        .ok_or(BlockProductionError::PoolBalanceNotFound(*pool_id))?;
+
+    Ok(pool_balance)
+}
+
+pub fn get_pool_balances_at_heights(
+    chainstate: &dyn ChainstateInterface,
+    min_height: BlockHeight,
+    max_height: BlockHeight,
+    pool_id: &PoolId,
+) -> Result<BTreeMap<BlockHeight, NonZeroPoolBalances>, BlockProductionError> {
+    let balances = chainstate
+        .get_stake_pool_balances_at_heights(&[*pool_id], min_height, max_height)
+        .map_err(|err| {
+            BlockProductionError::ChainstateError(consensus::ChainstateError::PoolBalanceReadError(
+                *pool_id,
+                err.to_string(),
+            ))
+        })?;
+
+    let balances = balances
+        .iter()
+        .filter_map(|(height, balances)| {
+            balances.get(pool_id).map(|balance| (*height, balance.clone()))
+        })
+        .collect::<BTreeMap<_, _>>();
+
+    Ok(balances)
```

Comment:
Would it make sense here to just return the iterator and let the client decide whether to collect it to a data structure, take just one element or what not. Since it produces elements lazily, you could even remove `max_height` and let the client decide when to stop iteration.